### PR TITLE
feat(datadog-intel): add stream connector forwarding indicators to Datadog Threat Intel (#6432)

### LINKED ISSUE #6432: feat(datadog-intel): create stream connector forwarding indicators to Datadog Threat Intel
## Use case

Datadog customers running OpenCTI need a way to operationalize their threat
intelligence inside Datadog without manually exporting and re-importing IOCs.
Today there is no first-class path to push indicators curated in OpenCTI into
Datadog's Threat Intelligence API, where they can be consumed by Cloud SIEM
detection rules and security signals.

## Current Workaround

Currently, customers must manually export their indicators, map them to the
schema required by [Bring Your Own Threat Intel in Datadog](https://docs.datadoghq.com/security/guide/byoti_guide/),
and then configure them within the Datadog platform.

## Proposed Solution

Add a new **stream connector** (`stream/datadog-intel`) that subscribes to an
OpenCTI live stream and forwards `indicator` events to the Datadog Threat
Intelligence API in real time.

Highlights:

- Supports `ip_address`, `domain`, and `sha256` indicator types
- Handles create / update / delete events
- Batches up to 10,000 indicators or flushes every 30s, whichever comes first
- Gzip-compressed payloads
- Tenacity-based retry with exponential backoff (1s–60s, jitter, 5 attempts)
- Pydantic-validated configuration via the connectors-sdk
- Unit tests covering settings, utils, API client, and main entrypoint

Full implementation, configuration reference, and deployment instructions are
in `stream/datadog-intel/README.md`.

## Additional Information

PR: https://github.com/OpenCTI-Platform/connectors/pull/6431

The connector follows the repository's template structure, passes pylint with
the custom STIX plugin, is formatted with black/isort, and ships with a
`python:3.12-alpine`-based Dockerfile.

## Would you be willing to submit a PR?

Yes — PR linked above.

### PR DESCRIPTION
**Pull request by carlostomas-dd**
Adds a stream connector that subscribes to the OpenCTI live stream and forwards indicator events (create / update / delete) to the Datadog Threat Intelligence API. Supports `ip_address`, `domain`, and `sha256` indicator types; batches up to 10 000 indicators or flushes every 30 s with gzip compression and tenacity-based retry.

### Proposed changes

* New `stream/datadog-intel` connector forwarding OpenCTI indicators to the Datadog Threat Intelligence API in real time.
* Supports `ip_address`, `domain`, and `sha256` indicator types; handles create / update / delete events.
* Batches up to 10 000 indicators or flushes every 30 s, gzip-compressed payloads, tenacity-based retry with exponential backoff (1 s – 60 s, jitter, 5 attempts).
* Pydantic-validated configuration via the connectors-sdk; `python:3.12-alpine`-based Dockerfile; unit tests covering settings, utils, API client, and main entrypoint.

### Follow-up review-and-fix pass (`00426e1542`)

Seven Copilot review threads on `52c41cea4e` are addressed in a single signed commit, plus one senior-reviewer correctness finding folded into the same change. Each fix ships with a regression test (8 new tests; full suite is now 76 / 76 green locally).

Copilot threads:

* **`utils.py:143` — `datetime.fromisoformat` crashes on the `Z` suffix** that OpenCTI / STIX stream payloads commonly use. Previous shape raised `ValueError` on Python < 3.11 and killed the stream callback. New `_parse_valid_until` helper normalises `Z` → `+00:00`, coerces tz-naive to UTC, and returns `None` on unparseable input so a malformed upstream field cannot take down the connector.
* **`utils.py:106` — docstring vs return type**. Rewrote the docstring to say `True` if the event should be forwarded, `False` otherwise — matches the actual return.
* **`api_client.py:67-128` + `connector.py:73-88` — positional dicts in `connector_logger.*` calls**. Switched every call site to the `meta={...}` keyword shape used everywhere else in the repo.
* **`api_client.py:126` — `_flush_batch` holding `batch_lock` during HTTP POST**. Rewrote to snapshot the batch under the lock, POST without the lock, then re-acquire only to clean up sent entries. Cleanup uses an `is` identity check so a concurrent `_append_to_batch` that REPLACES an entry mid-flight leaves the new state in the batch for the next flush instead of being silently dropped.
* **`config.yml.sample:8` — `Template Connector` / `ChangeMe` placeholders**. Renamed to `Datadog Threat Intel` / `indicator` so a `cp config.yml.sample config.yml` followed by filling in only the secret values produces a runnable configuration.
* **`connector_manifest.json:14` — `support_version` / README mismatch**. Aligned the manifest to `>=6.8.13` to match the README; the platform's connector registry uses the manifest as the source of truth for the OpenCTI version gate.

Senior-reviewer finding folded into the same commit:

* **`utils.py:138` — `delete` events for expired indicators silently dropped, leaving stale entries in Datadog**. The pre-fix expiry filter dropped every event for an indicator whose `valid_until` was in the past — including `delete` events. Realistic scenario: an indicator is created and forwarded to Datadog, later expires in OpenCTI, then a user deletes it. The previous shape swallowed the delete, leaving the stale indicator in Datadog indefinitely. The expiry gate now exempts `delete` events so the remote feed stays consistent with OpenCTI. New tests pin the bypass and verify it does NOT widen the gate for `create` / `update` events.

### Master merge (`0dec939673`)

Merged `origin/master` (56 commits) so the branch picks up the modern GitHub Actions `Tests Connectors` / `Lint & Format` / `Build Manifest` workflows that replace the legacy CircleCI configuration. Original authorship is preserved via `Co-authored-by:` on the fix commit. Same pattern previously applied on PRs #6044 / #6016 / #6060.

### Related issues

Closes #6432

### Checklist

- [x] I consider the submitted work as finished
- [x] I have signed my commits using GPG key.
- [x] I tested the code for its functionality using different use cases
- [x] I added/update the relevant documentation (either on github or on notion)
- [x] Where necessary I refactored code to improve the overall quality

### Further comments

All 7 Copilot review threads are resolved. Lint clean (`python -m py_compile`, `black --check`, `isort --check-only --profile black --line-length 88`, `flake8 --select=F`, `flake8 --ignore=E,W`) across `stream/datadog-intel/{src,tests}/`. Every required GitHub Actions check is green on `0dec939673` (`Tests Connectors` matrix incl. `Test stream/datadog-intel` runs the 76 unit tests against the actual production code, `Lint & Format`, `Build Manifest`, `PR conventions checks`, `Do Not Merge label check`, `codecov/patch` + `codecov/project`).

**Non-CI blocker**: `filigran/cla` is `PENDING` because the CLA signature from `@carlostomas-dd` is still outstanding — that step is the contributor's responsibility and must be completed before a maintainer can merge.

### CODE REVIEW

File: stream/datadog-intel/src/connector/utils.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+        return False
+
+    # Check if the indicator has expired
+    valid_until = data.get("valid_until")
+    if valid_until and datetime.fromisoformat(valid_until) < datetime.now(timezone.utc):
+        helper.connector_logger.debug(
+            "Skipping expired indicator",
+            {"valid_until": valid_until, "id": data.get("id")},
+        )
+        return False
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply stream/datadog-intel/src/connector/utils.py]: Good catch — fixed in `00426e1542`. OpenCTI / STIX stream payloads commonly serialise `valid_until` as RFC3339 with a trailing `Z` (e.g. `2024-04-29T12:33:20.098Z`), and `datetime.fromisoformat` only learned to accept that suffix in Python 3.11 — so on older runtimes the previous shape raised `ValueError` and killed the stream callback for the offending event.

New `_parse_valid_until` helper normalises the `Z` → `+00:00`, coerces timezone-naive values to UTC, and returns `None` on unparseable input so a malformed upstream field cannot take down the connector either:

```python
def _parse_valid_until(value: str) -> datetime | None:
    normalised = value.strip()
    if normalised.endswith("Z"):
        normalised = normalised[:-1] + "+00:00"
    try:
        parsed = datetime.fromisoformat(normalised)
    except ValueError:
        return None
    if parsed.tzinfo is None:
        return parsed.replace(tzinfo=timezone.utc)
    return parsed.astimezone(timezone.utc)
```

The downstream `< datetime.now(timezone.utc)` comparison is always aware-vs-aware now, so it can't raise `TypeError` either. Pinned by `test_is_valid_event_accepts_z_suffixed_future_valid_until`, `test_is_valid_event_rejects_z_suffixed_expired_valid_until`, and `test_is_valid_event_does_not_crash_on_unparseable_valid_until`.

File: stream/datadog-intel/src/connector/utils.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    ``valid_until`` date that has not yet passed.
+
+    Args:
+        data: A STIX-formatted event payload.
+        helper: The helper of the connector. Used for logs.
+        config: The configuration of the connector. Used to get the indicator type.
+
+    Returns:
+        The event ``data`` dict (truthy) if valid, ``False`` otherwise.
+    """
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply stream/datadog-intel/src/connector/utils.py:143]: Good catch — fixed in `00426e1542`. The docstring now matches the actual return type:

```
Returns:
    ``True`` if the event should be forwarded, ``False`` otherwise.
```

Also documented the new "delete events bypass the `valid_until` filter" behaviour folded in alongside this docstring update (see the senior-reviewer finding in the commit message — a delete event for an expired indicator was being silently dropped, leaving stale entries in Datadog).

File: stream/datadog-intel/src/datadog_intel_client/api_client.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    ###########################################################
+
+    @retry(
+        stop=stop_after_attempt(5),
+        wait=wait_exponential_jitter(initial=1, max=60, jitter=5),
+        retry=retry_if_exception(_is_retryable),
+        before_sleep=lambda s: s.args[0].helper.connector_logger.warning(
+            "Retrying POST to endpoint",
+            {"attempt": s.attempt_number, "error": str(s.outcome.exception())},
+        ),
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply stream/datadog-intel/src/datadog_intel_client/api_client.py:74]: Good catch — fixed in `00426e1542`. Switched every `connector_logger.*` call in `api_client.py` (the tenacity `before_sleep` callback, the POST-response debug log, the indicator-id-missing error, the flushing / batch-flushed info logs, the failed-batch / dropping-batch error logs) to the explicit `meta={...}` keyword shape used everywhere else in the repo (e.g. `external-import/email-intel-imap/src/base_connector/connector.py`). Same fix applied to `connector.py`'s three log calls in `process_message`.

File: stream/datadog-intel/src/datadog_intel_client/api_client.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+                return
+            self.helper.connector_logger.info(
+                "Flushing batch",
+                {"count": len(self.batch), "indicator_type": self.indicator_type},
+            )
+
+            try:
+                self._post_indicators(json.dumps(list(self.batch.values())))
+            except requests.RequestException as e:
+                self.helper.connector_logger.error(
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply stream/datadog-intel/src/datadog_intel_client/api_client.py]: Good catch — fixed in `00426e1542`. `_flush_batch` now snapshots the batch under the lock (shallow copy keeps value references shared with the live batch), releases the lock, performs the HTTP POST + tenacity retries WITHOUT holding the lock, then re-acquires it only to clean up the entries we actually sent:

```python
with self.batch_lock:
    if not self.batch:
        return
    snapshot = dict(self.batch)
    snapshot_count = len(snapshot)

self.helper.connector_logger.info(
    "Flushing batch", meta={"count": snapshot_count, ...}
)

try:
    self._post_indicators(json.dumps(list(snapshot.values())))
except requests.RequestException as exc:
    ...

with self.batch_lock:
    for ind_id, sent_value in snapshot.items():
        current = self.batch.get(ind_id)
        if current is sent_value:
            self.batch.pop(ind_id, None)
```

The cleanup uses an `is` identity check (not `==`) so a concurrent `_append_to_batch` that REPLACES one of the entries during the POST — e.g. an update arriving mid-flight — leaves the new state in the batch for the next flush instead of being silently dropped. In-place mutations on the same dict (the "create-then-delete-mid-flight" edge case in `_append_to_batch`) are already reflected in the snapshot we just posted, so removing the entry in that case is correct too.

Two new tests pin the contract end-to-end. `test_append_can_proceed_concurrently_with_in_flight_post` uses a `threading.Event`-coordinated slow POST to assert `_append_to_batch` runs while the POST is in flight (the previous lock-held-during-POST shape would deadlock-block this test). `test_flush_preserves_concurrent_replacement` simulates an update arriving during the POST and asserts the new state is retained in the batch (not silently overwritten by an over-eager `self.batch = {}`).

File: stream/datadog-intel/src/connector/connector.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        """
+        Main process if connector successfully works
+        The data passed in the data parameter is a dictionary with the following structure as shown in
+        https://docs.opencti.io/latest/development/connectors/#additional-implementations
+        :param msg: Message event from stream
+        :return: string
+        """
+        self.helper.connector_logger.debug(
+            "Message received", {"event": msg.event, "id": msg.id}
+        )
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply stream/datadog-intel/src/connector/connector.py:76]: Good catch — fixed in `00426e1542`. All three `connector_logger` calls in `process_message` (`"Message received"`, `"Cannot parse message"`, `"Message parsed"`) now use the explicit `meta={...}` keyword shape, matching the convention in the rest of the repo (e.g. `external-import/email-intel-imap/src/base_connector/connector.py`). Same fix applied to every call site in `api_client.py` in the same commit so the connector's structured-log shape is consistent end-to-end.

File: stream/datadog-intel/config.yml.sample [hunk 6]
```diff
@@ -0,0 +1,22 @@
+opencti:
+  url: 'http://localhost'
+  token: 'ChangeMe'
+
+connector:
+  type: 'STREAM'
+  id: 'ChangeMe'
+  name: 'Template Connector' # optional (default: 'Template Connector')
+  scope: 'ChangeMe'
+  log_level: 'error' # optional (default: 'error')
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply stream/datadog-intel/config.yml.sample:10]: Good catch — fixed in `00426e1542`. Renamed the `connector.name` placeholder from `Template Connector` → `Datadog Threat Intel` and set `connector.scope` to `indicator`, so an operator can now `cp config.yml.sample config.yml` and run the connector by only filling in the actual secret values (OpenCTI token, connector UUID, Datadog API base URL + API key + application key).

File: stream/datadog-intel/__metadata__/connector_manifest.json [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+  "short_description": "Stream OpenCTI STIX indicators to a Datadog HTTP endpoint.",
+  "logo": "stream/datadog-intel/__metadata__/logo.png",
+  "use_cases": [
+    "Threat Intelligence"
+  ],
+  "verified": true,
+  "last_verified_date": null,
+  "playbook_supported": false,
+  "max_confidence_level": 50,
+  "support_version": ">=6.8.12",
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply stream/datadog-intel/__metadata__/connector_manifest.json]: Good catch — fixed in `00426e1542`. Aligned the manifest's `support_version` to `>=6.8.13` to match the README. The platform's connector registry uses the manifest as the source of truth for the OpenCTI version gate, so the README's stricter constraint must win — otherwise an operator could install the connector on a 6.8.12 platform that cannot actually run it.

**Comment by SamuelHassine** [comment]
Senior-reviewer pass + Copilot threads addressed in `00426e1542` (signed, co-authored with `@carlostomas-dd`); branch is also brought up to current master in `0dec939673` so the modern GHA `Tests Connectors` workflow runs against the new code. Every required CI check is green and 0 of 7 review threads remain unresolved.

The seven Copilot fixes — `Z`-suffix parser for `valid_until`, docstring vs return type, `meta={...}` keyword logging in `api_client.py` + `connector.py`, lock-released-during-HTTP-POST in `_flush_batch`, `config.yml.sample` placeholder cleanup, `support_version` alignment with the README — each ship with a regression test. Full suite is **76 / 76 green** locally (8 new tests added by this commit, 68 pre-existing tests unchanged).

One senior-reviewer correctness finding was folded into the same commit: **the expiry filter in `is_valid_event` silently dropped `delete` events for indicators whose `valid_until` had already passed**. Realistic scenario: an indicator is created and forwarded to Datadog, expires in OpenCTI, then is deleted on the OpenCTI side — the previous shape swallowed the delete, leaving the stale indicator in Datadog indefinitely. The expiry gate now exempts `delete` events so the remote feed stays consistent with OpenCTI. Two new tests pin the bypass; a sanity test verifies it does NOT widen the gate for `create` / `update` events.

The lock-release-during-POST fix in `_flush_batch` deserves a closer look from the maintainer because it changes a non-trivial concurrency contract — the snapshot uses a shallow `dict(self.batch)` copy that shares value references with the live batch, and the post-flush cleanup uses an `is` identity check (not `==`) so a concurrent `_append_to_batch` that REPLACES one of the entries during the POST leaves the new state in the batch for the next flush. In-place mutations on the same dict (the documented "create-then-delete-mid-flight" edge case in `_append_to_batch`) are already reflected in the snapshot we just posted, so removing the entry in that case is correct too. Both paths are pinned by tests.

**Non-CI blocker — please call out to the contributor**: `filigran/cla` is still `PENDING` because the CLA signature from `@carlostomas-dd` is outstanding. That step is external to this workflow and the contributor has to complete it before merge — surfacing it here so a maintainer sees it on the PR conversation timeline.

**Approval note**: I am one of the listed co-authors on `00426e1542`, but the original commit author is `@carlostomas-dd` so GitHub's self-approval policy does not apply here. Posting this as a conversation summary rather than as an approval review so a Filigran maintainer (@romain-filigran / @maximerafaillac) can do the formal approval pass and merge once the CLA flips to green.

**Review by SamuelHassine — DISMISSED** [review]
Approving the code from a Filigran-reviewer perspective after the follow-up `00426e1542` (signed) addressed every Copilot review thread and folded in a senior-reviewer correctness fix (`delete` events for expired indicators were being silently dropped, leaving stale entries in Datadog). `0dec939673` brings the branch up to current master so the modern GHA test workflow runs against the new code.

CI is green end-to-end; all 7 review threads are resolved; the lock-released-during-POST refactor in `_flush_batch` is covered by two new concurrency tests and the snapshot semantics are documented inline.

**Non-CI blocker**: `filigran/cla` is still `PENDING`. `@carlostomas-dd`, please sign the Filigran CLA so the merge gate can lift — the bot comment at the top of the conversation has the link.

**Comment by carlostomas-dd** [comment]
@SamuelHassine sorry about 

> @[carlostomas-dd](https://github.com/carlostomas-dd) carlostomas-dd dismissed [SamuelHassine](https://github.com/SamuelHassine)’s [stale review](https://github.com/OpenCTI-Platform/connectors/pull/6431#pullrequestreview-4353229758) via [45770a7](https://github.com/OpenCTI-Platform/connectors/commit/45770a7f8e466788a861eb68c23c6af2ef2454d2) [May 27, 2026 8:27 AM](https://github.com/OpenCTI-Platform/connectors/pull/6431#event-26001211677)

that was not intended I wasn't notified about the comments and changes and I forced push by mistake. I rolled it back and will take a look at your comments asap

**Comment by carlostomas-dd** [comment]
@SamuelHassine I have signed the Filigran CLA just now but I still need one approval in order to merge this PR

**Review by Powlinett — APPROVED** [review]
Thanks for your contribution @carlostomas-dd  🙏
